common/tools/dev-tool/src/commands/run/build-test.ts

```diff
@@ -0,0 +1,209 @@
+  "browser-test": {
+    kind: "boolean",
+    default: true,
+    description: "build a bundle for browser testing",
+  },
+  "browser-config": {
+    kind: "string",
+    default: "tsconfig.browser.config.json",
+    description: "path to the browser config file",
+  },
+});
+
+export default leafCommand(commandInfo, async (options) => {
+  const browserTest = options["browser-test"];
+  const browserConfig = options["browser-config"];
+
+  const info = await resolveProject(process.cwd());
+
+  let moduleField = info.packageJson.module;
+  if (!moduleField) {
+    const defaultExport = info.packageJson.exports?.["."]?.import as any;
+    if (defaultExport) {
+      moduleField = defaultExport?.default;
+    }
+  }
+
+  if (!moduleField) {
+    log.error(info.name, "does not specify a `module` field or `exports` top level field.");
+    return false;
+  }
+
+  // Read the imports
+  const importMap = new Map<string, string>();
+  const importField = (info.packageJson as any)?.imports;
+  if (importField) {
+    const keys = Object.keys(importField);
+    for (const key of keys) {
+      importMap.set(key, importField[key]);
+    }
+  }
+
+  // Build the overrides - hard code to browser for now
+  const overrides = new Map<string, OverrideSet>();
+  overrides.set("esm", new OverrideSet("esm", "browser"));
+  const sources = new Set(getSources());
+  for (const file of sources) {
+    for (const override of overrides.values()) {
+      override.addOverride(file, sources);
+    }
+  }
+
+  for (const [name, pf] of overrides.entries()) {
+    if (pf.map.size === 0) {
+      overrides.delete(name);
+    }
+  }
+
+  if (browserTest) {
+    log.info(`Building for browser testing...`);
+    const esmMap = overrides.has("esm") ? overrides.get("esm")!.map : new Map<string, string>();
+    compileForEnvironment("browser", browserConfig, importMap, esmMap);
+  }
+
+  return true;
+});
+
+function compileForEnvironment(
+  type: string,
+  tsConfig: string,
+  importMap: Map<string, string>,
+  overrideMap: Map<string, string>,
+): boolean {
+  const tsconfigPath = path.join(process.cwd(), tsConfig);
+  const tsConfigFile = readFileSync(tsconfigPath, "utf8");
+  const tsConfigJSON = JSON.parse(tsConfigFile);
+  const outputPath = tsConfigJSON.compilerOptions.outDir;
+  if (!existsSync(tsconfigPath)) {
+    log.error(`TypeScript config ${tsConfig} does not exist`);
+    return false;
+  }
+
+  const browserTestPath = path.join(process.cwd(), outputPath);
+  if (!existsSync(browserTestPath)) {
+    mkdirSync(browserTestPath, { recursive: true });
+  }
+
+  // Create import map
+  const imports: Record<string, string> = {};
+  for (const [key, value] of importMap.entries()) {
+    imports[key] = value;
+  }
+
+  const packageJson = {
+    type: "module",
+    imports,
+  };
+  writeFileSync(path.join(browserTestPath, "package.json"), JSON.stringify(packageJson, null, 2));
+
+  const res = spawnSync(`tsc -p ${tsConfig}`, [], {
```

Comment:
this technically opens dev-tool up to injection attacks since `shell: true` is set, do we have another way of invoking external commands without worrying about arguments injecting shell commands?

Reply:
Fixed by adding validation to the tsconfig:

```
    if (!existsSync(browserConfig)) {
      log.error(`The file ${browserConfig} does not exist.`);
      return false;
    }

    const browserConfigStat = statSync(browserConfig);
    if (!browserConfigStat.isFile()) {
      log.error(`The file ${browserConfig} does not exist.`);
      return false;
    }
```

Reply:
I think I'm convinced that this does mitigate the shell exec issue because we're ensuring the string is a path to a file that actually exists on the system, but could we not just do the exec-form of spawn without the shell? What actually requires the shell here?